Entities/Characters/Scripts/RunnerDeath.as

```diff
@@ -40,12 +40,20 @@ f32 onHit(CBlob@ this, Vec2f worldPoint, Vec2f velocity, f32 damage, CBlob@ hitt
 			}
 		}
 
-		// add pickup attachment so we can pickup body
+		// add/convert pickup attachment so we can pickup body
 		CAttachment@ a = this.getAttachments();
 
 		if (a !is null)
 		{
-			AttachmentPoint@ ap = a.AddAttachmentPoint("PICKUP", false);
+			AttachmentPoint@ point = a.getAttachmentPointByName("PICKUP");
```

Comment:
This part of the logic seems to cause a change in the offset of picked up corpses:

Before:  
![before image](https://user-images.githubusercontent.com/8986152/216774143-3bfb4e45-83fa-41f5-adac-2f60e323a527.png)

After:  
![after image](https://user-images.githubusercontent.com/8986152/216774032-756e8576-1b95-41cb-9418-64a1c8a59869.png)

This is probably undesirable as it reduces the usefulness of corpses for hiding your team color.

Reply:
Is there a method to delete or move the offset an existing AttachmentPoint? 

I believe the issue with the previous code was that it was creating duplicate "PICKUP" AttachmentPoints, so the existing "socket" AttachmentPoint was throwing off this check: https://github.com/transhumandesign/kag-base/blob/8e936d2ebec1fa126213b6e8b185baca07a605ed/Entities/Characters/Knight/KnightLogic.as#L1754

But maybe I'm just misunderstanding how AddAttachmentPoint works. Is there any documentation for methods (with parameters) on these classes? There's nothing about it in [http://doc.kag2d.com/](http://doc.kag2d.com/)

Reply:
I figured out how to update the offset to `(0,0)`, but that _still_ had a regression with builder corpses because builders already have two attachment points defined in `Builder.cfg`, one as a socket and one as a plug - for building I guess. 
https://github.com/transhumandesign/kag-base/blob/8e936d2ebec1fa126213b6e8b185baca07a605ed/Entities/Characters/Builder/Builder.cfg#L209

That plug is used when you pick up the corpse, instead of the newly added one from `RunnerDeath.as`. And since the existing plug is offset vertically, it actually covers more of the player than a new `(0,0)` plug.

So for now, I'm "deleting" the existing pickup socket by re-naming it. Please let me know if there's a better way to remove it.

Reply:
Oh, I didn't realize why the double `PICKUP` attachment problem was even a thing.

I don't think you can remove attachment points. I'd be wary of any side effects of being able to remove an object like this (such as questionable indexing or sneaky dangling references). The engine code seems to index into the internal attachment list by index in a few places, so removing attachment points would seem risky to me either way.

The plug vs socket case is really annoying. It seems like a flaw in the original API that it is possible to query a single attachment point by name. Searching for `getAttachmentPointByName("PICKUP")` in Base reveals that this `KnightLogic.as` line you linked is seemingly the only one that might be looking for a socket/plug pickup attachment point on a player blob.

Using `getAttachmentPoints` and manually looking for a plug `PICKUP` point (while commenting why that is necessary) might be a cleaner way to achieve this.

Reply:
> Using `getAttachmentPoints` and manually looking for a plug `PICKUP` point (while commenting why that is necessary) might be a cleaner way to achieve this.

The trouble I'm getting with this approach is that a builder holding an item can't be slashed/jabbed. That's because `isAttachedToPoint()` takes a string name and not an `AttachmentPoint`, so I can't check if the target is specifically attached to the plug `PICKUP` point rather than the socket. Is there a better method for that?

And that takes me back to my earlier question: Is there a way to check the methods/properties of a class like `AttachmentPoint`? Do I need to use reflection (if that's even a thing)?

Reply:
> And that takes me back to my earlier question: Is there a way to check the methods/properties of a class like AttachmentPoint? Do I need to use reflection (if that's even a thing)?

`Manual/Interface/` in the KAG directory. For class methods, either in `Objects.txt` or `Objects/`. Those are only the function signatures though, there is no documentation (other than `doc.kag2d.com`).

Also, after looking stuff up, I noticed that there _is_ a method to `CAttachment` get an attachment point specifically of a desired socket/plug type:

```
	AttachmentPoint@ getAttachmentPoint(const string&in)
	AttachmentPoint@ getAttachmentPoint(const string&in, bool socket)
	AttachmentPoint@ getAttachmentPoint(const string&in, bool socket, bool occupied)
```

There is also `getAttachmentWithBlob(CBlob@ blob)`. So you could

1. Check that the hit candidate is a `PICKUP`, occupied plug
2. Check the name for the attachment with a common blob to ensure that its name is `PICKUP`

Sprinkled with some sanity null checks just in case.

Reply:
> 2. Check the name for the attachment with a common blob to ensure that its name is `PICKUP`

I don't understand what you mean by this.

Pushed a new version that uses `getAttachmentPoint(string, bool, bool)`, which seems to work. It seems like the PICKUP plug defined in Builder.cfg doesn't actually get occupied when you select blocks.

I also added a check for the `"player"` tag so that players carried by gregs can still be hit. (This was previously working by accident for the same reason that held corpses could get slashed.)